Import iNaturalist observations to Symbiota

mickley opened the pull request:
This adds iNaturalist import capability to Symbiota. Observations can be imported to a collection, and will show up around the portal, or can be used to move data into specimen records, print labels, etc. 

## Examples
<img width="1039" height="828" alt="Screenshot 2025-07-24 at 5 26 46 PM" src="https://github.com/user-attachments/assets/9df182eb-97b1-4e7a-9e97-ec82627c4c53" />
<img width="940" height="781" alt="Screenshot 2025-07-24 at 5 28 09 PM" src="https://github.com/user-attachments/assets/e061be56-97a7-4543-9607-599d3d755868" />

## Features:
- Allows searching/filtering and importing of iNaturalist observations
- Images can be imported, though the physical image files reside with iNaturalist, Symbiota only stores links.
- Imported records have a linked resource to iNaturalist as an externalOccurrence, with relationship type iNaturalistObservation. This custom type needs to be added to the ctcontrolvocabterm table. 
- Users must authenticate with iNaturalist and get an API token. This relaxes the API limits somewhat, and is preferred by iNat, they like to know who is doing what. 
- Optionally include elevation based on coordinates using Google's elevation API (or OpenElevation, but free version of this is rate- and request-limited to 1/second and 10,000/month). 
- Optionally, include associatedTaxa based on a radius search for other iNat observations from the same observer/date. 
- Optionally, site-level data can be set once for a whole group of observations: locality, habitat, associatedTaxa, and associatedCollectors. This overwrites any existing data in those fields. 
- Optionally, include data in custom observation fields of iNaturalist observations (which could be additional Darwin Core fields or anything else). 
- Optionally, add a link to the imported Symbiota record to the iNaturalist observation (e.g., if it represents a physical specimen). This uses the observation field "Link to Symbiota Record" on iNaturalist, but could be configured to something else, e.g., portal-specific. 
- Optionally update records already in the portal (similar to full upload for csv files). If turned off, records already present are not imported. 
- Saves search and import options



# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [ ] Hotfixes should be branched off of the `master` branch and PR'd using the **merge** option (not squashed) into the `hotfix` branch.
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
- [ ] There are no linter errors
- [ ] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
- [ ] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [ ] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
- [ ] Comment which GitHub issue(s), if any does this PR address
- [ ] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).

# Post-Approval

- [ ] It is the code author's responsibility to merge their own pull request after it has been approved
- [ ] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [ ] If this PR represents a merge into the `hotfix` branch, remember to use the **merge** option (i.e., no squash).
- [ ] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [ ] If this PR represents a merge from the `hotfix` branch into the `master` branch use the **squash & merge** option
  - [ ] a subsequent PR from `master` into `Development` should be made with the **merge** option (i.e., no squash).
  - [ ] **Immediately** delete the `hotfix` branch and create a new `hotfix` branch
  - [ ] increment the Symbiota version number in the symbase.php file and commit to the `hotfix` branch.
- [ ] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [ ] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!

mickley committed:
Bugfix: zeros are now shown when reviewing uploaded data

There was a bug where if a field was zero, it would be shown as a blank field in the upload reviewer. Additionally, if all records were null or zero for a field, then the field would not be shown. This caused issues with variables such as elevation that could have values of zero.

mickley committed:
Adds copyright info and rights URL for individual media on public record displays

Previously, if the media copyright field was set, the rights for the record itself were set to match the media. However, this is not great form, as both the data and the media can and often do have different rights. For example, many collections license data as CC-BY-NC, and images as CC-BY-NC-SA. Additionally, either the data or the media may have more restrictive licences than the other component. For example, the data may be licensed under a creative commons license, while the media may be public domain (thus making the entire record public domain under the current paradigm). Or the media  could be copyrighted by a different individual and used with permission, while the data copyright is  more permissive.

A better way of handling this is to allow both sets of rights to be displayed, so this adds the image copyright (if set) to be displayed below the media.

mickley committed:
Fixes compatibility issues with PHP < 8.0 for classes/Media.php

See: https://github.com/Symbiota/Symbiota/issues/2416

mickley committed:
Allows importing records for Personal Observation Management Collections

This was already possible for admins, but not for users. However, it's a feature that is useful for collectors pulling in their data. For now, only full text file import is enabled, no skeletal import or import profiles.

mickley committed:
Updates associatedOccurrence JSON import code to use new omoccurassociations field names

- identifier -> objectID
- type -> associationType

mickley committed:
Adds iNaturalist import capability for collections.

This lets iNaturalist observations be imported as either specimens (in a specimen collection) or as unvouchered observations in an observation collection.

mickley commented:
## Some implementation notes:
- Some mostly unrelated fixes are included in separate commits
  - Compatibility issues with PHP < 8.0 in Media.php, See: https://github.com/Symbiota/Symbiota/issues/2416
  - Allows for importing records for Personal Observation Collections (currently only full text file import, and iNat import in the subsequent commit)
  - Adds copyright info and rights URL for individual media on public record displays. This allows for different rights for data and images. 
  - A bugfix in the uploaded data reviewer page (uploadreviewer.php) where data with zeros were not displayed.
- The interface is using language files, and I've populated them, but I haven't translated the Spanish and French ones. 
- This uses a javascript wrapper library to interface with iNaturalist that I wrote: https://github.com/mickley/iNatJS/. You're welcome to use it for other things. It's using jsdelivr to serve the js file tracking the main branch, but you could also include the file in Symbiota or use a release if you like. 
- Licensing: Currently, cc-by, cc-by-nc, and cc0 are allowed to be imported for both data and images, unless the authenticated user is the observer. This list of licenses can be changed in collections.inat.js
  - These are the same licenses that iNaturalist sends to GBIF for research-grade observations. iNaturalist says that they've received legal advice that observation data isn't copyrightable, but they don't push it. 
- There are some portal-manager configuration settings at the top of collections.inat.js

mickley commented:
Oh, and I haven't gotten to Symbiota 3.3 yet. So this is built on 3.2.

mickley committed:
Adds an initial determination history entry for iNaturalist observation imports.
This is the same as what is done for file uploads

mickley committed:
Bugfix: Correctly handles empty fields when creating an initial determination on import
When records are imported from files (or iNaturalist), they create an initial determination based on the sciname, identifiedBy, and dateIdentified fields. The query checks for null fields and defaults to things like "Unknown", or "s.d", but if the fields are not null, but empty, that isn't caught.

It's possible for a field to be non-null but empty here with imports when the field is included in the import and has data for one record, but not others. I don't think this is an issue in the other places where determinations are automatically created.

Atticus29 commented:
@MuchQuak @mickley this is slated for 3.5 (sorry for how long it's taken, James).

mickley commented:
> @MuchQuak @mickley this is slated for 3.5 (sorry for how long it's taken, James).

Great to hear, no worries!

mickley committed:
Fixes bug where re-importing data loses basisOfRecord

Previously, when importing a record that is already present in the database, if basisOfRecord wasn't explicitly specified, the previous basisOfRecord would be lost. Now, the existing basisOfRecord will be used if one isn't specified in the import.

This applies to re-importing iNaturalist observations, but also to other import use cases.

mickley committed:
Fixes iNat import bug where an observation has no annotations array
